gordo_components/server/utils.py

```diff
@@ -168,6 +132,53 @@ def parse_iso_datetime(datetime_str: str) -> datetime:
     return parsed_date
 
 
+def _df_from_dict_or_respond(
+    data: Union[dict, list], expected_columns: List[str]
+) -> Union[Response, pd.DataFrame]:
+    """
+    Convert a dict or list of dicts into a :class:`pandas.core.DataFrame`.
+    Setting the column names to ``expected_columns`` if not already labeled and
+    the length of the columns match the length of the epxected columns.
+
+    If it fails, it will return an instance of :class:`flask.wrappers.Response`
+
+    Parameters
+    ----------
+    data: Union[dict, list]
+        Data to convert into a dataframe with :meth:`pandas.core.DataFrame.from_dict`
+    expected_columns: List[str]
+        List of expected column names to give if the dataframe does not consist of them
+        but the number of columns matches ``len(expected_columns)``
+
+    Returns
+    -------
+    Union[flask.wrappers.Response, pandas.core.DataFrame]
+    """
+    df = pd.DataFrame.from_dict(data)
+
+    if not all(col in df.columns for col in expected_columns):
+
+        # If the length doesn't mach, then we can't reliably determine what data we have hre.
+        if len(df.columns) != len(expected_columns):
+            msg = dict(
+                message=f"Unexpected features:"
+                f" was expecting {expected_columns} length of {len(expected_columns)}, "
+                f"but got {df.columns} length of {len(df.columns)}"
+            )
+            return make_response((jsonify(msg), 400))
```

Comment:
I hate this! Is it not better to raise e.g. ValueError, and then let the caller make the stupid response object if it wants?

Reply:
Hate is a strong word. 

Basically this is only taking place in `extract_X_y` decorator, so if it cannot do that, it will return an error before the decorated method has the opportunity to execute, because it simply _can't_ without such data being extracted. 

with that said, I'm ok raising an error in this function, but in `extract_X_y` returning the response. Heck, even could return a fancy `enum`if you like. :+1:

Reply:
> with that said, I'm ok raising an error in this function, but in extract_X_y returning the response. 
What a lovely idea!
> Heck, even could return a fancy enumif you like. 
No thanks:-)

Reply:
Although, the message won't be as informative as it won't have access to the previously built dataframe occuring in this function. 

```rust
let df = match _df_or_resp_from_dict(data) {
    Dataframe(df) => df,
    Response(resp) => return resp
}
```

Reply:
You can just put the message in the ValueError and use it in the response.